File: service/benchmark.py
"""轻量压测脚本 —— 模拟 N 个并发用户打 /chat。

用法：
  uv run -m service.server &          # 先起服务
  uv run service/benchmark.py --users 200 --requests 1000

依赖：pip install httpx
不依赖重量级框架，便于在本地快速验证并发模型（单进程 asyncio 客户端）。
"""

from __future__ import annotations

import argparse
import asyncio
import time

import httpx


async def _one_user(client: httpx.AsyncClient, base: str, user_id: str, conv_id: str, n: int):
    ok = 0
    for i in range(n):
        try:
            r = await client.post(
                f"{base}/chat",
                json={
                    "user_id": user_id,
                    "conversation_id": conv_id,
                    "message": f"我想买冰箱，家里 {i % 5 + 2} 口人，预算 5000。",
                },
                timeout=60,
            )
            if r.status_code == 200:
                ok += 1
        except Exception:
            pass
    return ok


async def main(users: int, requests: int):
    base = "http://localhost:8000"
    per = max(1, requests // users)
    start = time.perf_counter()
    async with httpx.AsyncClient() as client:
        tasks = [
            _one_user(client, base, f"u{u}", f"c{u}", per)
            for u in range(users)
        ]
        results = await asyncio.gather(*tasks)
    elapsed = time.perf_counter() - start
    total = sum(results)
    print(f"users={users} total_requests={total} elapsed={elapsed:.1f}s "
          f"rps={total / elapsed:.1f} success_rate={total / (users * per):.1%}")


if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("--users", type=int, default=200)
    ap.add_argument("--requests", type=int, default=1000)
    args = ap.parse_args()
    asyncio.run(main(args.users, args.requests))

File: presale-demo/src/benchmark.ts
import { argv } from "node:process";

/**
 * 轻量压测脚本 —— 模拟 N 个并发用户打 /chat。
 * 用法：pnpm bench --users 200 --requests 1000
 * 依赖 axios/fetch（Node 内置 fetch 即可，无需额外依赖）。
 */

function parseArgs(): { users: number; requests: number } {
  const users = Number(argv.find((a, i) => a === "--users" && argv[i + 1]) ?? 200);
  const requests = Number(argv.find((a, i) => a === "--requests" && argv[i + 1]) ?? 1000);
  return { users: users || 200, requests: requests || 1000 };
}

async function oneUser(base: string, userId: string, convId: string, n: number): Promise<number> {
  let ok = 0;
  for (let i = 0; i < n; i++) {
    try {
      const res = await fetch(`${base}/chat`, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({
          user_id: userId,
          conversation_id: convId,
          message: `我想买冰箱，家里 ${i % 5 + 2} 口人，预算 5000。`,
        }),
      });
      if (res.status === 200) ok++;
    } catch {
      // 忽略网络错误，计入失败
    }
  }
  return ok;
}

async function main() {
  const { users, requests } = parseArgs();
  const base = `http://localhost:${process.env.PORT ?? 8000}`;
  const per = Math.max(1, Math.floor(requests / users));
  const start = Date.now();

  const tasks = Array.from({ length: users }, (_, u) =>
    oneUser(base, `u${u}`, `c${u}`, per),
  );
  const results = await Promise.all(tasks);

  const elapsed = (Date.now() - start) / 1000;
  const total = results.reduce((a, b) => a + b, 0);
  console.log(
    `users=${users} total_requests=${total} elapsed=${elapsed.toFixed(1)}s ` +
      `rps=${(total / elapsed).toFixed(1)} success_rate=${((total / (users * per)) * 100).toFixed(1)}%`,
  );
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
